Regex allocation and comm reduction

aconsroe-hpe opened the pull request:
While we landed better support for multilocale regex by implementing `chpl__serialize` in #18678, we weren't actually packing the pattern string/bytes inline. This PR fixes that and results in reduced comm counts when the pattern is small enough to fit inside the string/bytes own inline buffer.

We can now also turn many of the heap allocations into stack allocations because the execution of the matching will now always happen on `here`.

Tests cover comm counts for serialization, allocation counts for local `search` and `matches` (where we expect to see our reduced allocations), and a general test of the new behavior for local+remote regexes.

* [x] Full paratest pass

ronawho commented on modules/standard/Regex.chpl:
```diff
@@ -833,27 +837,35 @@ record regex {
+    if home != here {
+      regexCopy = this;
+      localRegex = regexCopy._regex;
     }
```
Note that this adds an extra remote GET when the `home != here`. This isn't a case we care about optimizing a lot, but seems like we could avoid that with something like:

```chpl
    var regexCopy:regex(exprType);
    if home != here then regexCopy = this;
    const localRegex = if home != here then regexCopy._regex else _regex;
```

aconsroe-hpe replied:
Do you mean the GET for transferring the regex? I do see one less `chpl_gen_comm_get` with your version, but that is removing the `get` for the first line `var localRegex = _regex;`. So I am inclined to use your version, though I'm still figuring out how/why its different

ronawho replied:
When `home != here`, previously,  `var localRegex = _regex` would do a GET, and then `localRegex` gets overridden by `localRegex = regexCopy._regex`, so it was a wasted GET.

In the pattern I suggested I don't think there's ever that extra GET. We may still be generating a call to `chpl_get_comm_get`, but I'm pretty sure there won't be a network remote GET at runtime, though it could be worth double checking the comm counts.

ronawho commented on test/regex/allocationCounts.good:
```diff
@@ -0,0 +1,18 @@
+Iter Matches no capture
+
+Iter Matches with capture
+
```
Any reason to skip having a test for iter split?

aconsroe-hpe replied:
Only because we don't expect to see anything interesting

ronawho commented on modules/standard/Regex.chpl:
```diff
@@ -870,30 +882,34 @@ record regex {
-    on this.home {
-      matches = _ddata_allocate(qio_regex_string_piece_t, nmatches);
-    }
+    matches = _ddata_allocate(qio_regex_string_piece_t, nmatches);
```
Shouldn't impact performance, but just to get rid of uses of _ddata, I think we should switch this to c_malloc/c_free (or c_calloc if we care about zero initing this)

ronawho commented on modules/standard/Regex.chpl:
```diff
@@ -586,10 +586,13 @@ proc bytes.this(m:regexMatch) {
 record chpl_serializeHelper {
   type exprType;
-  var pattern:exprType;
+  var pattern:(new exprType()).chpl__serialize().type;
```
I'm not sure the current state of this, but at least in the past we didn't DCE expressions that were only used to compute a type. Can you verify this isn't resulting in an allocation?

aconsroe-hpe replied:
Hmm I'm pretty sure it doesn't but not 100% sure I'm looking in all the right places. The type is definitely resolved at compile time 
```C
typedef struct chpl_chpl_serializeHelper_string_s {
  chpl____serializeHelper pattern;
  qio_regex_options_t options;
} chpl_serializeHelper_string;
```
and I don't see any calls to `chpl__serialize*` in eg module init code or anywhere unexpected

mppf replied:
We have trouble with DCE of expressions computing a type when the type has a runtime component (arrays, domains)

ronawho reviewed (approved):
Left some small feedback inline, but overall looks good. 

An upstream Arkouda PR was just merged that impacts their regex performance, so let's wait a day or 2 before merging this to isolate that perf impact.

aconsroe-hpe commented:
Okay sounds good. Replaced `_ddata` and added to `serializationCommCounts` to also run a `match` when the regex is remote. I didn't look into this deeply, but maybe we can use it as a baseline and future work could continue to improve it. (But also happy to integrate more changes if you have ideas right now)

ronawho commented:
> but maybe we can use it as a baseline and future work could continue to improve it. (But also happy to integrate more changes if you have ideas right now)

Yeah, I'm on board with that generally, but thought this was a simple fix worth getting into this one.

aconsroe-hpe commented:
Just to clarify I _did_ use your suggested change on the `localRegex` business, thanks for that! I was referring to something that would be even better than that (which I don't know how feasible that is)

mppf commented on modules/standard/Regex.chpl:
```diff
@@ -586,10 +586,13 @@ proc bytes.this(m:regexMatch) {
 record chpl_serializeHelper {
   type exprType;
-  var pattern:exprType;
+  var pattern:(new exprType()).chpl__serialize().type;
```
IMO this would be better written as

``` chapel
// using a function for this makes the field declaration less complex and allows reuse
private proc getSerializedType(type exprType) type {
  var tmpStr: exprType; // default init of a string variable is more common than `new string()`
  var ser = tmpStr.chpl__serialize();
  return ser.type;
}
record chpl_serializeHelper {
  type exprType;
  var pattern:exprType;
  var pattern: getSerializedType(exprType);
  ...
```

aconsroe-hpe replied:
much less spicy, but looks good :)

mppf commented on modules/standard/Regex.chpl:
```diff
@@ -636,7 +639,7 @@ record regex {
   pragma "no doc"
   proc _serialize() {
-    var pattern: exprType;
+    var pattern: (new exprType()).chpl__serialize().type;
```
let's use a `getSerializedType` here

mppf commented on modules/standard/Regex.chpl:
```diff
@@ -833,27 +837,29 @@ record regex {
   {
+    /* This pattern is a bit ugly, but we'd like to avoid
+    unnecessarily copying `this` when local. And we need to maintain
+    the lifetime of the copied regex when remote. */
```
I think it'd be clearer with "maintain the lifetime of the copied regex when remote" -> "deallocate the temporary copy when remote"

mppf replied:
Also, I think the way I can see to make this less ugly is to add `regex.localize()` similar to `string.localize()` that creates an aliased copy when it is local. (But of course that complicates the regex memory management). I don't think you need to do that in this PR but you could put a comment describing the alternative.

mppf replied:
Right, but saying something like "deallocate in the right place" vs the "lifetime" part would be nice. "lifetime" makes me think about a bunch of other issues (that certainly encompass deallocation).

aconsroe-hpe replied:
If "lifetime" encompasses "deallocate" when you read it, then I'm not inclined to change it

ronawho replied:
For the `localize` wrapper, did we try something like:


```chpl
inline proc regex.localize() {
  var ret:regex(exprType);
  if _local || this.home == here {
    ret._regex = this._regex;
  } else {
    ret = this;
  }
  return ret;
}
```

aconsroe-hpe replied:
I think I considered that and the issue is that when `ret` gets `deinit` in the enclosing scope (like `_search_match`), it would do a refcount decrement (with no matching increment)

ronawho replied:
Ah, yeah that's right. I think we could avoid that by adding a "refCount" or "isBorrowed" or something bool and set that in localize and avoid the refcount decrement if it's set. I'd be fine doing that as a followup cleanup though.

And actually I think string has to have similar functionality and it's probably called `isOwned` there so I'd just mimic that.

mppf replied:
Right. I think we should leave that for another PR though. Also arguably the string localize shouldn't return a string but rather a string view of sorts. That way we can have the lifetime checker handle the checking that the string view doesn't outlive the string.

mppf commented on modules/standard/Regex.chpl:
```diff
@@ -833,27 +837,29 @@ record regex {
+    endpos = pos + text.numBytes;
+
+    param nmatches = if has_captures then 1 + captures.size else 1;
+    var matches = new c_array(qio_regex_string_piece_t, nmatches);
```
Could alternatively be written as `var matches: c_array(qio_regex_string_piece_t, nmatches);` which I guess is more what I would expect since you're really default initializing (the arguments are type information). Not a big deal though.

aconsroe-hpe replied:
Yeah looking at other modules, they all use the form you suggested, changing now
